### DISCUSSION TITLE
behind-the-scenes-of-collection-expressions-part-5-adding-support-for-collection-expressions-to-your-own-types/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#426] [Announcements]
# behind-the-scenes-of-collection-expressions-part-5-adding-support-for-collection-expressions-to-your-own-types/

In this final post of the series I show how to add support for collection expressions to your own types that don't support them automatically

https://andrewlock.net/behind-the-scenes-of-collection-expressions-part-5-adding-support-for-collection-expressions-to-your-own-types/

**Comment by MarsonShine** [comment]
nice article.
section in "Adding [CollectionBuilder] to interfaces", i have some trouble: 
```c#
[CollectionBuilder(typeof(MyCollectionBuilder), nameof(MyCollectionBuilder.Create))]
public interface IMyCollection<T> : IEnumerable<T>
{
}

public class MyCollectionBuilder
{
	public static MyBetterCollection<T> Create<T>(ReadOnlySpan<T> values) => new(values);
}

public class MyImplementCollection<T>(T[] values) : IMyCollection<T>
{
	public IEnumerator<T> GetEnumerator() => values.AsEnumerable().GetEnumerator();

	IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}
```

finally, `IMyCollection<int> c5 = [1, 2, 3, 4]` was wrong: CS9187: Could not find an accessible method with the expected signature: a static method with a single parameter of type ReadOnlySpan<T>; and correct return type.
am i missing some details?

**Reply by wekempf** [reply]
The error message I got was better than you said here, as it stated the return type was expected to be IMyCollection<T>. However, that wasn't completely accurate, as I was able to return a MyImplementCollection<T>. My guess, without having your full source, is that you don't have a MyBetterCollection<T> and intended to return a MyImplementCollection<T>.

Full source of what worked for me. Note that there were some other errors I had to fix in the Create method.

```csharp
[CollectionBuilder(typeof(MyCollectionBuilder), nameof(MyCollectionBuilder.Create))]
public interface IMyCollection<T> : IEnumerable<T>
{
}

public class MyCollectionBuilder
{
	public static MyImplementCollection<T> Create<T>(ReadOnlySpan<T> values) => new MyImplementCollection<T>(values.ToArray());
}

public class MyImplementCollection<T>(T[] values) : IMyCollection<T>
{
	public IEnumerator<T> GetEnumerator() => values.AsEnumerable().GetEnumerator();

	IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}
```
